icechunk/src/session.rs

```diff
@@ -0,0 +1,1065 @@
+        let read_only = self.read_only();
+
+        Store::new(self.clone(), StoreOptions::default(), read_only)
+    }
+
+    pub async fn get_node(&self, path: &Path) -> RepositoryResult<NodeSnapshot> {
+        let guard = self.change_set.read().await;
+        get_node(self.storage.as_ref(), &self.snapshot_id, &guard, path).await
+    }
+
+    pub async fn get_array(&self, path: &Path) -> RepositoryResult<NodeSnapshot> {
+        let guard = self.change_set.read().await;
+        get_array(self.storage.as_ref(), &self.snapshot_id, &guard, path).await
+    }
+
+    pub async fn get_group(&self, path: &Path) -> RepositoryResult<NodeSnapshot> {
+        let guard = self.change_set.read().await;
+        get_group(self.storage.as_ref(), &self.snapshot_id, &guard, path).await
+    }
+
+    pub async fn get_chunk_ref(
+        &self,
+        path: &Path,
+        coords: &ChunkIndices,
+    ) -> RepositoryResult<Option<ChunkPayload>> {
+        let guard = self.change_set.read().await;
+        get_chunk_ref(self.storage.as_ref(), &self.snapshot_id, &guard, path, coords)
+            .await
+    }
+
+    /// Get a future that reads the the payload of a chunk from object store
+    ///
+    /// This function doesn't return [`Bytes`] directly to avoid locking the ref to self longer
+    /// than needed. We want the bytes to be pulled from object store without holding a ref to the
+    /// [`Repository`], that way, writes can happen concurrently.
+    ///
+    /// The result of calling this function is None, if the chunk reference is not present in the
+    /// repository, or a [`Future`] that will fetch the bytes, possibly failing.
+    ///
+    /// Example usage:
+    /// ```ignore
+    /// get_chunk(
+    ///     ds.get_chunk_reader(
+    ///         &path,
+    ///         &ChunkIndices(vec![0, 0, 0]),
+    ///         &ByteRange::ALL,
+    ///     )
+    ///     .await
+    ///     .unwrap(),
+    /// ).await?
+    /// ```
+    ///
+    /// The helper function [`get_chunk`] manages the pattern matching of the result and returns
+    /// the bytes.
+    pub async fn get_chunk_reader(
+        &self,
+        path: &Path,
+        coords: &ChunkIndices,
+        byte_range: &ByteRange,
+    ) -> RepositoryResult<
+        Option<Pin<Box<dyn Future<Output = RepositoryResult<Bytes>> + Send>>>,
+    > {
+        let guard = self.change_set.read().await;
+        get_chunk_reader(
+            &self.storage,
+            &self.virtual_resolver,
+            &self.snapshot_id,
+            &guard,
+            path,
+            coords,
+            byte_range,
+        )
+        .await
+    }
+
+    pub async fn get_old_chunk(
+        &self,
+        node: NodeId,
+        manifests: &[ManifestRef],
+        coords: &ChunkIndices,
+    ) -> RepositoryResult<Option<ChunkPayload>> {
+        // FIXME: use manifest extents
+        get_old_chunk(self.storage.as_ref(), node, manifests, coords).await
+    }
+
+    pub async fn list_nodes(
```

Comment:
@paraseba any ideas on how to fix this? This currently is not correct because the iterator grabs the lifetime of the RWLockGuard which is temporary, so returning an iterator does not compile